results/pkg/api/server/server_test.go

```diff
@@ -374,24 +417,70 @@ func TestListResults(t *testing.T) {
 			name:         "test if field is case sensitive",
 			filter:       `taskrun.MetaData.name=="notaskrun"`,
 			expectStatus: codes.InvalidArgument,
+		}, {
+			name:          "test query with pagesize and pagetoken",
+			filter:        `taskrun.api_version=="v1beta1"`,
+			expect:        []*pb.Result{gotResults[2], gotResults[3]},
+			expectStatus:  codes.OK,
+			pageSize:      2,
+			pageToken:     encodeOffset(1),
+			nextPageToken: encodeOffset(4),
+		}, {
+			name:          "test query with pagesize and pagetoken, no more next page",
+			filter:        `taskrun.api_version=="v1beta1"`,
+			expect:        []*pb.Result{gotResults[2], gotResults[3], gotResults[8]},
+			expectStatus:  codes.OK,
+			pageSize:      3,
+			pageToken:     encodeOffset(1),
+			nextPageToken: "",
+		}, {
+			name:          "test query with pagesize and pagetoken, the last page, got insufficient pages",
+			filter:        `taskrun.api_version=="v1beta1"`,
+			expect:        []*pb.Result{gotResults[3], gotResults[8]},
+			expectStatus:  codes.OK,
+			pageSize:      3,
+			pageToken:     encodeOffset(3),
+			nextPageToken: "",
+		}, {
+			name:          "test query with overstepped offset",
+			filter:        "",
+			expect:        nil,
+			expectStatus:  codes.OK,
+			pageSize:      3,
+			pageToken:     encodeOffset(10),
+			nextPageToken: "",
 		},
 	}
 	// Test if we can find inserted taskruns
 	for _, tc := range tt {
 		t.Run(tc.name, func(t *testing.T) {
 			res, err := srv.ListResultsResult(ctx, &pb.ListResultsRequest{
-				Filter: tc.filter,
+				Filter:    tc.filter,
+				PageSize:  tc.pageSize,
+				PageToken: tc.pageToken,
 			})
 			if tc.expectStatus != status.Code(err) {
 				t.Fatalf("failed on test %v: %v", tc.name, err)
 			}
 			gotList := res.GetResults()
+			nextPageToken := res.GetNextPageToken()
+			var expectedNextOffset int64 = 0
+			var gotNextOffset int64 = 0
+			if tc.nextPageToken != "" {
+				expectedNextOffset, _ = decodeToken(tc.nextPageToken)
```

Comment:
We should have a separate test for Token encoding/decoding. This test isn't sufficient on its own, since if decodeToken always returned a static value this would still pass.

Reply:
Thanks for the great question!

I did this just to make it easy for human to understand what behind the token. I should just compare the `tc.nextPageToken` and `nextPageToken` :(

As for the seperate test for Token encoding/decoding, I guess the tests  above about `pagination`--different pageToken and pageSize leads to different results and nextPageToken--could be enough?  At least returning a static value will certainly fail on thoses tests.